File: examples/grid_holy_grail.rs
// This creates a so-called "holy grail" layout using the CSS Grid layout algorithm
// See: https://en.wikipedia.org/wiki/Holy_grail_(web_design)

// NOTE: This example requires the `grid` feature flag to be enabled.

#[cfg(not(feature = "grid"))]
fn main() {
    println!("Error: this example requires the 'grid' feature to be enabled");
    println!("Try:");
    println!("    cargo run --example grid_holy_grail --features grid")
}

#[cfg(feature = "grid")]
fn default<T: Default>() -> T {
    Default::default()
}

#[cfg(feature = "grid")]
fn main() -> Result<(), taffy::error::TaffyError> {
    use taffy::prelude::*;

    let mut taffy = Taffy::new();

    // Setup the grid
    let root_style = Style {
        display: Display::Grid,
        size: Size { width: points(800.0), height: points(600.0) },
        grid_template_columns: vec![points(250.0), fr(1.0), points(250.0)],
        grid_template_rows: vec![points(150.0), fr(1.0), points(150.0)],
        ..default()
    };

    // Define the child nodes
    let header = taffy.new_leaf(Style { grid_row: line(1), grid_column: span(3), ..default() })?;
    let left_sidebar = taffy.new_leaf(Style { grid_row: line(2), grid_column: line(1), ..default() })?;
    let content_area = taffy.new_leaf(Style { grid_row: line(2), grid_column: line(2), ..default() })?;
    let right_sidebar = taffy.new_leaf(Style { grid_row: line(2), grid_column: line(3), ..default() })?;
    let footer = taffy.new_leaf(Style { grid_row: line(3), grid_column: span(3), ..default() })?;

    // Create the container with the children
    let root = taffy.new_with_children(root_style, &[header, left_sidebar, content_area, right_sidebar, footer])?;

    // Compute layout and print result
    taffy.compute_layout(root, Size { width: points(800.0), height: points(600.0) })?;
    taffy::debug::print_tree(&taffy, root);

    Ok(())
}

File: bindings/java/java/src/examples/java/com/dioxuslabs/taffy/examples/GridHolyGrail.java
package com.dioxuslabs.taffy.examples;

import com.dioxuslabs.taffy.Dimension;
import com.dioxuslabs.taffy.Display;
import com.dioxuslabs.taffy.GridPlacement;
import com.dioxuslabs.taffy.GridTrack;
import com.dioxuslabs.taffy.Style;
import com.dioxuslabs.taffy.TaffyTree;

/**
 * Java port of {@code examples/grid_holy_grail.rs} — the classic 3-column,
 * 3-row holy-grail layout: header, left sidebar, content, right sidebar,
 * footer.
 */
public final class GridHolyGrail {

    private GridHolyGrail() {}

    public static void main(String[] args) {
        try (TaffyTree tree = new TaffyTree()) {
            long header = tree.newLeaf(Style.builder()
                    .gridRow(GridPlacement.line(1), GridPlacement.AUTO)
                    .gridColumn(GridPlacement.span(3), GridPlacement.AUTO));

            long leftSidebar = tree.newLeaf(Style.builder()
                    .gridRow(GridPlacement.line(2), GridPlacement.AUTO)
                    .gridColumn(GridPlacement.line(1), GridPlacement.AUTO));

            long contentArea = tree.newLeaf(Style.builder()
                    .gridRow(GridPlacement.line(2), GridPlacement.AUTO)
                    .gridColumn(GridPlacement.line(2), GridPlacement.AUTO));

            long rightSidebar = tree.newLeaf(Style.builder()
                    .gridRow(GridPlacement.line(2), GridPlacement.AUTO)
                    .gridColumn(GridPlacement.line(3), GridPlacement.AUTO));

            long footer = tree.newLeaf(Style.builder()
                    .gridRow(GridPlacement.line(3), GridPlacement.AUTO)
                    .gridColumn(GridPlacement.span(3), GridPlacement.AUTO));

            long root = tree.newWithChildren(Style.builder()
                            .display(Display.GRID)
                            .size(Dimension.length(800f), Dimension.length(600f))
                            .gridTemplateColumns(GridTrack.length(250f), GridTrack.fr(1f), GridTrack.length(250f))
                            .gridTemplateRows(GridTrack.length(150f), GridTrack.fr(1f), GridTrack.length(150f)),
                    header, leftSidebar, contentArea, rightSidebar, footer);

            tree.computeLayout(root, 800f, 600f);

            System.out.println("root:         " + tree.layout(root));
            System.out.println("header:       " + tree.layout(header));
            System.out.println("leftSidebar:  " + tree.layout(leftSidebar));
            System.out.println("contentArea:  " + tree.layout(contentArea));
            System.out.println("rightSidebar: " + tree.layout(rightSidebar));
            System.out.println("footer:       " + tree.layout(footer));
        }
    }
}
